### ISSUE TITLE
Conversion trait implementation issue related to `ndarray`

### ISSUE THREAD

**Issue by Ilia-Kosenkov** [extendr/extendr#500]
There is an issue with `TryFrom` implemented for `ArrayView1`,  but not for `ArrayView2`.

``` r
rextendr::rust_source(
    code = "#[extendr(use_try_from = true)] fn this_works(_x : ArrayView2<Rint>) {}",
    extendr_deps = list(
        `extendr-api` = list(
            git = "https://github.com/extendr/extendr",
            features = array("ndarray")
        )
    )
)
#> ℹ build directory: 'C:/Users/.../AppData/Local/Temp/RtmpAXOM9P/file3f586e29777a'
#> ✔ Writing 'C:/Users/.../AppData/Local/Temp/RtmpAXOM9P/file3f586e29777a/target/extendr_wrappers.R'

rextendr::rust_source(
    code = "#[extendr(use_try_from = true)] fn this_doesnt(_x : ArrayView1<Rint>) {}",
    extendr_deps = list(
        `extendr-api` = list(
            git = "https://github.com/extendr/extendr",
            features = array("ndarray")
        )
    )
)
#> ℹ build directory:
#> 'C:/Users/.../AppData/Local/Temp/RtmpAXOM9P/file3f586e29777a'
#> Error in `invoke_cargo()`:
#> ! Rust code could not be compiled successfully. Aborting.
#> ✖ error[E0277]: the trait bound `ArrayBase<ViewRepr<&extendr_api::prelude::Rint>, Dim<[usize; 1]>>: From<Robj>` is not satisfied
#>  --> src\lib.rs:2:1
#>   |
#> 2 | #[extendr(use_try_from = true)] fn this_doesnt(_x : ArrayView1<Rint>) {}
#>   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `From<Robj>` is not implemented for `ArrayBase<ViewRepr<&extendr_api::prelude::Rint>, Dim<[usize; 1]>>`
#>   |
#>   = help: the following other types implement trait `From<T>`:
#>             <ArrayBase<CowRepr<'a, A>, D> as From<&'a ArrayBase<S, D>>>
#>             <ArrayBase<CowRepr<'a, A>, D> as From<ArrayBase<OwnedRepr<A>, D>>>
#>             <ArrayBase<CowRepr<'a, A>, D> as From<ArrayBase<ViewRepr<&'a A>, D>>>
#>             <ArrayBase<CowRepr<'a, A>, Dim<[usize; 1]>> as From<&'a Slice>>
#>             <ArrayBase<OwnedArcRepr<A>, D> as From<ArrayBase<OwnedRepr<A>, D>>>
#>             <ArrayBase<OwnedRepr<A>, Dim<[usize; 2]>> as From<Vec<V>>>
#>             <ArrayBase<OwnedRepr<A>, Dim<[usize; 3]>> as From<Vec<V>>>
#>             <ArrayBase<S, Dim<[usize; 1]>> as From<Box<[A]>>>
#>           and 5 others
#>   = note: required for `Robj` to implement `Into<ArrayBase<ViewRepr<&extendr_api::prelude::Rint>, Dim<[usize; 1]>>>`
#>   = note: required for `ArrayBase<ViewRepr<&extendr_api::prelude::Rint>, Dim<[usize; 1]>>` to implement `TryFrom<Robj>`
#>   = note: required for `Robj` to implement `extendr_api::TryInto<ArrayBase<ViewRepr<&extendr_api::prelude::Rint>, Dim<[usize; 1]>>>`
#>   = note: this error originates in the attribute macro `extendr` (in Nightly builds, run with -Z macro-backtrace for more info)
#> ✖ error: aborting due to previous error

#> Backtrace:
#>     ▆
#>  1. └─rextendr::rust_source(...)
#>  2.   └─rextendr:::invoke_cargo(...)
#>  3.     └─rextendr:::check_cargo_output(...)
#>  4.       └─rextendr:::ui_throw(...)
#>  5.         └─rlang::abort(message, class = "rextendr_error", call = call)
```

<sup>Created on 2023-03-04 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by CGMossa** [comment]
Alright; Maybe this macro is the problem: https://extendr.github.io/extendr/src/extendr_api/robj_ndarray.rs.html#97
It is not exactly like the macro for view 2. It isn't making both impls in the view 1.

**Comment by CGMossa** [comment]
Try to use this branch: https://github.com/extendr/extendr/pull/501

**Comment by Ilia-Kosenkov** [comment]
Now it is a lifetime issue

<details>
<summary> Lengthy output </summary>

``` r
rextendr::rust_source(
    code = "#[extendr(use_try_from = true)] fn this_works(_x : ArrayView2<Rint>) {}",
    extendr_deps = list(
        `extendr-api` = list(
            git = "https://github.com/extendr/extendr",
            branch = "fix_ndarray_view_1",
            features = array("ndarray")
        )
    )
)
#> ℹ build directory:
#> 'C:/Users/.../AppData/Local/Temp/RtmpAHNM5D/file8342c211fb2'
#> Error in `invoke_cargo()`:
#> ! Rust code could not be compiled successfully. Aborting.
#> ✖ error[E0597]: `robj` does not live long enough
#>    --> C:\Users\...\.cargo\git\checkouts\extendr-569c02e1a164cdf5\487258c\extendr-api\src\robj_ndarray.rs:100:32
#>     |
#> 96  |         impl<'a> TryFrom<Robj> for ArrayView1<'a, $type> {
#>     |              -- lifetime `'a` defined here
#> ...
#> 99  |             fn try_from(robj: Robj) -> Result<Self> {
#>     |                         ---- binding `robj` declared here
#> 100 |                 Self::try_from(&robj)
#>     |                 ---------------^^^^^-
#>     |                 |              |
#>     |                 |              borrowed value does not live long enough
#>     |                 argument requires that `robj` is borrowed for `'a`
#> 101 |             }
#>     |             - `robj` dropped here while still borrowed
#> ...
#> 142 | make_array_view_1!(Rbool, Error::ExpectedLogical);
#>     | ------------------------------------------------- in this macro invocation
#>     |
#>     = note: this error originates in the macro `make_array_view_1` (in Nightly builds, run with -Z macro-backtrace for more info)
#> ✖ error[E0597]: `robj` does not live long enough
#>    --> C:\Users\...\.cargo\git\checkouts\extendr-569c02e1a164cdf5\487258c\extendr-api\src\robj_ndarray.rs:100:32
#>     |
#> 96  |         impl<'a> TryFrom<Robj> for ArrayView1<'a, $type> {
#>     |              -- lifetime `'a` defined here
#> ...
#> 99  |             fn try_from(robj: Robj) -> Result<Self> {
#>     |                         ---- binding `robj` declared here
#> 100 |                 Self::try_from(&robj)
#>     |                 ---------------^^^^^-
#>     |                 |              |
#>     |                 |              borrowed value does not live long enough
#>     |                 argument requires that `robj` is borrowed for `'a`
#> 101 |             }
#>     |             - `robj` dropped here while still borrowed
#> ...
#> 143 | make_array_view_1!(Rint, Error::ExpectedInteger);
#>     | ------------------------------------------------ in this macro invocation
#>     |
#>     = note: this error originates in the macro `make_array_view_1` (in Nightly builds, run with -Z macro-backtrace for more info)
#> ✖ error[E0597]: `robj` does not live long enough
#>    --> C:\Users\...\.cargo\git\checkouts\extendr-569c02e1a164cdf5\487258c\extendr-api\src\robj_ndarray.rs:100:32
#>     |
#> 96  |         impl<'a> TryFrom<Robj> for ArrayView1<'a, $type> {
#>     |              -- lifetime `'a` defined here
#> ...
#> 99  |             fn try_from(robj: Robj) -> Result<Self> {
#>     |                         ---- binding `robj` declared here
#> 100 |                 Self::try_from(&robj)
#>     |                 ---------------^^^^^-
#>     |                 |              |
#>     |                 |              borrowed value does not live long enough
#>     |                 argument requires that `robj` is borrowed for `'a`
#> 101 |             }
#>     |             - `robj` dropped here while still borrowed
#> ...
#> 144 | make_array_view_1!(i32, Error::ExpectedInteger);
#>     | ----------------------------------------------- in this macro invocation
#>     |
#>     = note: this error originates in the macro `make_array_view_1` (in Nightly builds, run with -Z macro-backtrace for more info)
#> ✖ error[E0597]: `robj` does not live long enough
#>    --> C:\Users\...\.cargo\git\checkouts\extendr-569c02e1a164cdf5\487258c\extendr-api\src\robj_ndarray.rs:100:32
#>     |
#> 96  |         impl<'a> TryFrom<Robj> for ArrayView1<'a, $type> {
#>     |              -- lifetime `'a` defined here
#> ...
#> 99  |             fn try_from(robj: Robj) -> Result<Self> {
#>     |                         ---- binding `robj` declared here
#> 100 |                 Self::try_from(&robj)
#>     |                 ---------------^^^^^-
#>     |                 |              |
#>     |                 |              borrowed value does not live long enough
#>     |                 argument requires that `robj` is borrowed for `'a`
#> 101 |             }
#>     |             - `robj` dropped here while still borrowed
#> ...
#> 145 | make_array_view_1!(u32, Error::ExpectedInteger);
#>     | ----------------------------------------------- in this macro invocation
#>     |
#>     = note: this error originates in the macro `make_array_view_1` (in Nightly builds, run with -Z macro-backtrace for more info)
#> ✖ error[E0597]: `robj` does not live long enough
#>    --> C:\Users\...\.cargo\git\checkouts\extendr-569c02e1a164cdf5\487258c\extendr-api\src\robj_ndarray.rs:100:32
#>     |
#> 96  |         impl<'a> TryFrom<Robj> for ArrayView1<'a, $type> {
#>     |              -- lifetime `'a` defined here
#> ...
#> 99  |             fn try_from(robj: Robj) -> Result<Self> {
#>     |                         ---- binding `robj` declared here
#> 100 |                 Self::try_from(&robj)
#>     |                 ---------------^^^^^-
#>     |                 |              |
#>     |                 |              borrowed value does not live long enough
#>     |                 argument requires that `robj` is borrowed for `'a`
#> 101 |             }
#>     |             - `robj` dropped here while still borrowed
#> ...
#> 146 | make_array_view_1!(Rfloat, Error::ExpectedReal);
#>     | ----------------------------------------------- in this macro invocation
#>     |
#>     = note: this error originates in the macro `make_array_view_1` (in Nightly builds, run with -Z macro-backtrace for more info)
#> ✖ error[E0597]: `robj` does not live long enough
#>    --> C:\Users\...\.cargo\git\checkouts\extendr-569c02e1a164cdf5\487258c\extendr-api\src\robj_ndarray.rs:100:32
#>     |
#> 96  |         impl<'a> TryFrom<Robj> for ArrayView1<'a, $type> {
#>     |              -- lifetime `'a` defined here
#> ...
#> 99  |             fn try_from(robj: Robj) -> Result<Self> {
#>     |                         ---- binding `robj` declared here
#> 100 |                 Self::try_from(&robj)
#>     |                 ---------------^^^^^-
#>     |                 |              |
#>     |                 |              borrowed value does not live long enough
#>     |                 argument requires that `robj` is borrowed for `'a`
#> 101 |             }
#>     |             - `robj` dropped here while still borrowed
#> ...
#> 147 | make_array_view_1!(f64, Error::ExpectedReal);
#>     | -------------------------------------------- in this macro invocation
#>     |
#>     = note: this error originates in the macro `make_array_view_1` (in Nightly builds, run with -Z macro-backtrace for more info)
#> ✖ error[E0597]: `robj` does not live long enough
#>    --> C:\Users\...\.cargo\git\checkouts\extendr-569c02e1a164cdf5\487258c\extendr-api\src\robj_ndarray.rs:100:32
#>     |
#> 96  |         impl<'a> TryFrom<Robj> for ArrayView1<'a, $type> {
#>     |              -- lifetime `'a` defined here
#> ...
#> 99  |             fn try_from(robj: Robj) -> Result<Self> {
#>     |                         ---- binding `robj` declared here
#> 100 |                 Self::try_from(&robj)
#>     |                 ---------------^^^^^-
#>     |                 |              |
#>     |                 |              borrowed value does not live long enough
#>     |                 argument requires that `robj` is borrowed for `'a`
#> 101 |             }
#>     |             - `robj` dropped here while still borrowed
#> ...
#> 148 | make_array_view_1!(Rcplx, Error::ExpectedComplex);
#>     | ------------------------------------------------- in this macro invocation
#>     |
#>     = note: this error originates in the macro `make_array_view_1` (in Nightly builds, run with -Z macro-backtrace for more info)
#> ℹ 3 compiler messages not shown.

#> Backtrace:
#>     ▆
#>  1. └─rextendr::rust_source(...)
#>  2.   └─rextendr:::invoke_cargo(...)
#>  3.     └─rextendr:::check_cargo_output(...)
#>  4.       └─rextendr:::ui_throw(...)
#>  5.         └─rlang::abort(message, class = "rextendr_error", call = call)

rextendr::rust_source(
    code = "#[extendr(use_try_from = true)] fn this_doesnt(_x : ArrayView1<Rint>) {}",
    extendr_deps = list(
        `extendr-api` = list(
            git = "https://github.com/extendr/extendr",
            branch = "fix_ndarray_view_1",
            features = array("ndarray")
        )
    )
)
#> ℹ build directory:
#> 'C:/Users/.../AppData/Local/Temp/RtmpAHNM5D/file8342c211fb2'
#> Error in `invoke_cargo()`:
#> ! Rust code could not be compiled successfully. Aborting.
#> ✖ error[E0597]: `robj` does not live long enough
#>    --> C:\Users\...\.cargo\git\checkouts\extendr-569c02e1a164cdf5\487258c\extendr-api\src\robj_ndarray.rs:100:32
#>     |
#> 96  |         impl<'a> TryFrom<Robj> for ArrayView1<'a, $type> {
#>     |              -- lifetime `'a` defined here
#> ...
#> 99  |             fn try_from(robj: Robj) -> Result<Self> {
#>     |                         ---- binding `robj` declared here
#> 100 |                 Self::try_from(&robj)
#>     |                 ---------------^^^^^-
#>     |                 |              |
#>     |                 |              borrowed value does not live long enough
#>     |                 argument requires that `robj` is borrowed for `'a`
#> 101 |             }
#>     |             - `robj` dropped here while still borrowed
#> ...
#> 142 | make_array_view_1!(Rbool, Error::ExpectedLogical);
#>     | ------------------------------------------------- in this macro invocation
#>     |
#>     = note: this error originates in the macro `make_array_view_1` (in Nightly builds, run with -Z macro-backtrace for more info)
#> ✖ error[E0597]: `robj` does not live long enough
#>    --> C:\Users\...\.cargo\git\checkouts\extendr-569c02e1a164cdf5\487258c\extendr-api\src\robj_ndarray.rs:100:32
#>     |
#> 96  |         impl<'a> TryFrom<Robj> for ArrayView1<'a, $type> {
#>     |              -- lifetime `'a` defined here
#> ...
#> 99  |             fn try_from(robj: Robj) -> Result<Self> {
#>     |                         ---- binding `robj` declared here
#> 100 |                 Self::try_from(&robj)
#>     |                 ---------------^^^^^-
#>     |                 |              |
#>     |                 |              borrowed value does not live long enough
#>     |                 argument requires that `robj` is borrowed for `'a`
#> 101 |             }
#>     |             - `robj` dropped here while still borrowed
#> ...
#> 143 | make_array_view_1!(Rint, Error::ExpectedInteger);
#>     | ------------------------------------------------ in this macro invocation
#>     |
#>     = note: this error originates in the macro `make_array_view_1` (in Nightly builds, run with -Z macro-backtrace for more info)
#> ✖ error[E0597]: `robj` does not live long enough
#>    --> C:\Users\...\.cargo\git\checkouts\extendr-569c02e1a164cdf5\487258c\extendr-api\src\robj_ndarray.rs:100:32
#>     |
#> 96  |         impl<'a> TryFrom<Robj> for ArrayView1<'a, $type> {
#>     |              -- lifetime `'a` defined here
#> ...
#> 99  |             fn try_from(robj: Robj) -> Result<Self> {
#>     |                         ---- binding `robj` declared here
#> 100 |                 Self::try_from(&robj)
#>     |                 ---------------^^^^^-
#>     |                 |              |
#>     |                 |              borrowed value does not live long enough
#>     |                 argument requires that `robj` is borrowed for `'a`
#> 101 |             }
#>     |             - `robj` dropped here while still borrowed
#> ...
#> 144 | make_array_view_1!(i32, Error::ExpectedInteger);
#>     | ----------------------------------------------- in this macro invocation
#>     |
#>     = note: this error originates in the macro `make_array_view_1` (in Nightly builds, run with -Z macro-backtrace for more info)
#> ✖ error[E0597]: `robj` does not live long enough
#>    --> C:\Users\...\.cargo\git\checkouts\extendr-569c02e1a164cdf5\487258c\extendr-api\src\robj_ndarray.rs:100:32
#>     |
#> 96  |         impl<'a> TryFrom<Robj> for ArrayView1<'a, $type> {
#>     |              -- lifetime `'a` defined here
#> ...
#> 99  |             fn try_from(robj: Robj) -> Result<Self> {
#>     |                         ---- binding `robj` declared here
#> 100 |                 Self::try_from(&robj)
#>     |                 ---------------^^^^^-
#>     |                 |              |
#>     |                 |              borrowed value does not live long enough
#>     |                 argument requires that `robj` is borrowed for `'a`
#> 101 |             }
#>     |             - `robj` dropped here while still borrowed
#> ...
#> 145 | make_array_view_1!(u32, Error::ExpectedInteger);
#>     | ----------------------------------------------- in this macro invocation
#>     |
#>     = note: this error originates in the macro `make_array_view_1` (in Nightly builds, run with -Z macro-backtrace for more info)
#> ✖ error[E0597]: `robj` does not live long enough
#>    --> C:\Users\...\.cargo\git\checkouts\extendr-569c02e1a164cdf5\487258c\extendr-api\src\robj_ndarray.rs:100:32
#>     |
#> 96  |         impl<'a> TryFrom<Robj> for ArrayView1<'a, $type> {
#>     |              -- lifetime `'a` defined here
#> ...
#> 99  |             fn try_from(robj: Robj) -> Result<Self> {
#>     |                         ---- binding `robj` declared here
#> 100 |                 Self::try_from(&robj)
#>     |                 ---------------^^^^^-
#>     |                 |              |
#>     |                 |              borrowed value does not live long enough
#>     |                 argument requires that `robj` is borrowed for `'a`
#> 101 |             }
#>     |             - `robj` dropped here while still borrowed
#> ...
#> 146 | make_array_view_1!(Rfloat, Error::ExpectedReal);
#>     | ----------------------------------------------- in this macro invocation
#>     |
#>     = note: this error originates in the macro `make_array_view_1` (in Nightly builds, run with -Z macro-backtrace for more info)
#> ✖ error[E0597]: `robj` does not live long enough
#>    --> C:\Users\...\.cargo\git\checkouts\extendr-569c02e1a164cdf5\487258c\extendr-api\src\robj_ndarray.rs:100:32
#>     |
#> 96  |         impl<'a> TryFrom<Robj> for ArrayView1<'a, $type> {
#>     |              -- lifetime `'a` defined here
#> ...
#> 99  |             fn try_from(robj: Robj) -> Result<Self> {
#>     |                         ---- binding `robj` declared here
#> 100 |                 Self::try_from(&robj)
#>     |                 ---------------^^^^^-
#>     |                 |              |
#>     |                 |              borrowed value does not live long enough
#>     |                 argument requires that `robj` is borrowed for `'a`
#> 101 |             }
#>     |             - `robj` dropped here while still borrowed
#> ...
#> 147 | make_array_view_1!(f64, Error::ExpectedReal);
#>     | -------------------------------------------- in this macro invocation
#>     |
#>     = note: this error originates in the macro `make_array_view_1` (in Nightly builds, run with -Z macro-backtrace for more info)
#> ✖ error[E0597]: `robj` does not live long enough
#>    --> C:\Users\...\.cargo\git\checkouts\extendr-569c02e1a164cdf5\487258c\extendr-api\src\robj_ndarray.rs:100:32
#>     |
#> 96  |         impl<'a> TryFrom<Robj> for ArrayView1<'a, $type> {
#>     |              -- lifetime `'a` defined here
#> ...
#> 99  |             fn try_from(robj: Robj) -> Result<Self> {
#>     |                         ---- binding `robj` declared here
#> 100 |                 Self::try_from(&robj)
#>     |                 ---------------^^^^^-
#>     |                 |              |
#>     |                 |              borrowed value does not live long enough
#>     |                 argument requires that `robj` is borrowed for `'a`
#> 101 |             }
#>     |             - `robj` dropped here while still borrowed
#> ...
#> 148 | make_array_view_1!(Rcplx, Error::ExpectedComplex);
#>     | ------------------------------------------------- in this macro invocation
#>     |
#>     = note: this error originates in the macro `make_array_view_1` (in Nightly builds, run with -Z macro-backtrace for more info)
#> ℹ 3 compiler messages not shown.

#> Backtrace:
#>     ▆
#>  1. └─rextendr::rust_source(...)
#>  2.   └─rextendr:::invoke_cargo(...)
#>  3.     └─rextendr:::check_cargo_output(...)
#>  4.       └─rextendr:::ui_throw(...)
#>  5.         └─rlang::abort(message, class = "rextendr_error", call = call)
```

<sup>Created on 2023-03-04 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>
</details>

**Comment by CGMossa** [comment]
I forgot that you need to do `cargo test --features ndarray`. Sorry about that.
I've pushed the lifetime fix.
I've also ran this
```r
rextendr::rust_source(
    code = "#[extendr(use_try_from = true)] fn this_doesnt(_x : ArrayView1<Rint>) {}",
    extendr_deps = list(
        `extendr-api` = list(
            git = "https://github.com/extendr/extendr",
            branch = "fix_ndarray_view_1",
            features = array("ndarray")
        )
    )
)
```
and it succeeded.

**Comment by Ilia-Kosenkov** [comment]
Works for me. Let's wait for CI to finish.

**Comment by CGMossa** [comment]
Fixed in #501
